vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -417,49 +479,110 @@ and
-
+                              FSharpConstants.FSharpLanguageName, 
+                              projectDisplayName, 
+                              projectFileName, 
+                              projectGuid, 
+                              hierarchy, 
+                              null, 
+                              errorReporter) :?> AbstractProject
+                | _ as p -> p
+
+            match project :> obj with
+            | :? IWorkspaceProjectContext as projectContext ->
                 this.SyncProject(project, projectContext, site, workspace, forceUpdate=false, userOpName=userOpName)
-                site.AdviseProjectSiteChanges(FSharpConstants.FSharpLanguageServiceCallbackName, 
-                                              AdviseProjectSiteChanges(fun () -> this.SyncProject(project, projectContext, site, workspace, forceUpdate=true, userOpName="AdviseProjectSiteChanges."+userOpName)))
-                site.AdviseProjectSiteClosed(FSharpConstants.FSharpLanguageServiceCallbackName, 
-                                             AdviseProjectSiteChanges(fun () -> 
+                site.AdviseProjectSiteChanges(FSharpConstants.FSharpLanguageServiceCallbackName,
+                                                AdviseProjectSiteChanges(fun () -> this.SyncProject(project, projectContext, site, workspace, forceUpdate=true, userOpName="AdviseProjectSiteChanges."+userOpName)))
+                site.AdviseProjectSiteClosed(FSharpConstants.FSharpLanguageServiceCallbackName,
+                                                AdviseProjectSiteChanges(fun () ->
                                                 projectInfoManager.ClearInfoForProject(project.Id)
                                                 optionsAssociation.Remove(projectContext) |> ignore
                                                 project.Disconnect()))
-                for referencedSite in ProjectSitesAndFiles.GetReferencedProjectSites (site, this.SystemServiceProvider) do
-                    let referencedProjectId = setup referencedSite                    
-                    project.AddProjectReference(ProjectReference referencedProjectId)
+            | _ -> ()
+            for referencedSite in ProjectSitesAndFiles.GetReferencedProjectSites (site, this.SystemServiceProvider) do
+                let referencedProjectId = setup referencedSite
+                project.AddProjectReference(ProjectReference referencedProjectId)
 
             projectId
         setup (siteProvider.GetProjectSite()) |> ignore
 
+    member private __.ProvideProjectSiteProvider(workspace:Workspace, project:Project) =
+        let visualStudioWorkspace = workspace :?> VisualStudioWorkspace
+        let hier = visualStudioWorkspace.GetHierarchy(project.Id)
+        {new IProvideProjectSite with
+             member this.GetProjectSite() =
+                 let compileItems () = projectInfoManager.GetSourceFiles(project.FilePath)
```

Comment:
I believe there's no need to define all these little functions, nor is it clearer - just put the implementations in the object expression